src/nvc.rs

```diff
@@ -2,6 +2,100 @@ use instruction::*;
 use interconnect::*;
 
 use std::collections::HashSet;
+use std::fmt;
+
+#[derive(Copy, Clone)]
+pub struct NvcCacheEntry {
+    pub tag: u32,
+    pub base_addr: u32,
+    pub subblock_valid: [bool; 2],
+}
+
+impl NvcCacheEntry {
+    fn new() -> NvcCacheEntry {
+        NvcCacheEntry {
+          tag: 0,
+          base_addr: 0,
+          subblock_valid: [false; 2],
+        }
+    }
+}
+
+impl fmt::Display for NvcCacheEntry {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "base_addr: 0x{:08x}, tag: 0x{:08x}, sb1 valid: {:5}, sb2 valid: {:5}", self.base_addr, self.tag, self.subblock_valid[0], self.subblock_valid[1])
+    }
+}
+
+pub struct NvcCache {
+    hits: usize,
+    misses: usize,
+    cache_entries: [NvcCacheEntry; 128],
+}
+
+impl NvcCache {
+    pub fn new() -> NvcCache {
+        NvcCache {
+          hits: 0,
+          misses: 0,
+          cache_entries: [NvcCacheEntry::new(); 128],
+        }
+    }
+
+    pub fn clear_entries(&mut self, entry_start: usize, count: usize) {
+        if entry_start >= 128 {
+            return;
+        }
+        let mut entry_count = count;
+
+        if entry_count > 128 {
+            entry_count = 128;
+        }
+
+        if entry_start + entry_count > 128 {
+            entry_count = 128 - entry_start;
+        }
+
+        for x in 0..entry_count {
+            self.cache_entries[entry_start + x].tag = 0;
+            self.cache_entries[entry_start + x].subblock_valid = [false; 2];
+         }
+    }
+
+   // if there is a cache miss the addr will be added to the cache
+   pub fn lookup(&mut self, addr: u32) -> bool {
+        let byte_offset = (addr & 0x07) as usize;
+        let entry_index = ((addr >> 3) & 0x7f) as usize;
+        let tag = addr >> 10;
+        let subblock = if byte_offset > 3 { 1 } else { 0 };
+
+        if self.cache_entries[entry_index].tag == tag {
+            if self.cache_entries[entry_index].subblock_valid[subblock] {
+                self.hits += 1;
+                return true;
+            }
+            self.cache_entries[entry_index].subblock_valid[subblock] = true;
+            self.misses += 1;
+            return false;
+        } else {
+            self.cache_entries[entry_index].tag = tag;
+            self.cache_entries[entry_index].subblock_valid = [false; 2];
+            self.cache_entries[entry_index].subblock_valid[subblock] = true;
+            self.cache_entries[entry_index].base_addr = addr & 0xfffffff8;
+            self.misses += 1;
+            return false;
+        }
+    }
+
+    pub fn get_entry(&self, entry_index: usize) -> NvcCacheEntry {
```

Comment:
Rust naming convention is `entry` for a getter, not `get_entry` (though a setter would still be `set_entry`) so I think that name would be preferred.